Initial CMake test support for Proxygen

MrAnno opened the pull request:
ProxygenTest.cmake initializes CTest and enables unit tests based on the
value of `BUILD_TESTING` (provided by CTest). This option is `OFF` by
default.

The module provides a function called `proxygen_add_test()`, which makes
it possible to declare tests without boilerplates:

```
proxygen_add_test(TARGET target
                  [SOURCES src1...]
                  [DEPENDS dep1...]
                  [INCLUDES inc1...]
                  [EXTRA_ARGS arg1...]
                  [WORKING_DIRECTORY dir]
                  [PREFIX prefix]
)
```

----

3 tests are left out:
- proxygen/httpserver/samples/echo/test
- HTTPServerTest.cpp (`undefined reference to folly::ssl::SSLServerOptions::kCipherList`)
- HTTPSessionAcceptorTest.cpp (`undefined reference to folly::ssl::SSLServerOptions::kCipherList`)

Note:
`GMOCK_METHOD1_(, noexcept, ...` does not work with GMock v1.8.1.

lnicco commented:
thank you for this PR!
We are reviewing internally and yes, there are problem with the tests you highlighted with ubuntu > 18.04 
I am assuming that is the distribution are you using?

MrAnno commented:
@lnicco I'm using Arch Linux, but I tried to be general enough in a hope that my patchset will work with Facebook's internal `fbcode_builder`s and Ubuntu-based `docker_builder`s as well.

I'm going to try modifying the proxygen `fbcode_builder` specs to use CMake, so I can test the pull request in your build environment.

MrAnno commented:
@lnicco Sorry for the delay. One last commit has been added to follow the changes of 154287a0a18c22c8674dc923b3d5bb8991cfd27e.

My attempts on building my branch with `fbcode_builder` were not very successful:
- I've modified the builder config to use `fb_github_cmake_install` with `BUILD_TESTING` enabled.
- `cmake_minimum_required(VERSION 3.10)` is too new for Xenial, so I've created a `--os-image ubuntu:18.04`, `--gcc-version 7` builder
  - Changing the min version to 3.5 is another option, but `gtest_add_tests` is not available in CMake 3.5, so something like this will be required: https://github.com/facebook/folly/blob/2bd32e7bd3527e2fccd2819fd90677a477d5d0e1/CMake/FollyFunctions.cmake#L265-L281

- Fixed a few Python 3 incompatibilities in `make_docker_context.py`

Unfortunately, this wasn't enough. It seems folly and wangle propagate `gflags_shared` as `INTERFACE_LINK_LIBRARIES` (`/home/install/lib/cmake/wangle/wangle-targets.cmake`), which does not exist in the builder environment.

It is also a problem that `specs/gmock.py` specifies `release-1.8.1`. If I'm not mistaken, I can't just change that tag for proxygen, because everything under `fbcode_builder` is shared between projects (folly, wangle, etc.) AFAIK.
